File: Top_100_Liked_Questions/62._Unique_Paths/solution.py
from math import factorial

class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return factorial(m+n-2) // (factorial(m-1) * factorial(n-1))
        
    
    def uniquePaths2(self, m: int, n: int) -> int:
        mem = [[1] * n for _ in range(m)]
        
        for i in range(1,m):
            for j in range(1,n):
                mem[i][j] = mem[i-1][j] + mem[i][j-1]
        
        return mem[-1][-1]

File: Top_100_Liked_Questions/62._Unique_Paths/solution.java
class Solution {
    public int uniquePaths(int m, int n) {
        int[][] mem = new int[n][m];

        for (int row=n-1; row>-1; row--) {
            for (int col=m-1; col>-1; col--) {
                if (row == n-1 && col == m-1)
                    mem[row][col] = 1;

                else if (row == n-1)
                    mem[row][col] = mem[row][col+1];

                else if (col == m-1)
                    mem[row][col] = mem[row+1][col];

                else
                    mem[row][col] = mem[row][col+1] + mem[row+1][col];
            }
        }

        return mem[0][0];
    }
}

class Solution2 {
    public int uniquePaths(int m, int n) {
        // uniquePaths(x,y) = uniquePaths(x+1,y) + uniquePaths(x,y+1)
        if (m == 1 && n == 1)
            return 1;

        else if (m == 1)
            return this.uniquePaths(m, n-1);

        else if (n == 1)
            return this.uniquePaths(m-1, n);

        else
            return this.uniquePaths(m-1, n) + uniquePaths(m, n-1);
    }
}